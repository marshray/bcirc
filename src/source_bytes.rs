// Copyright 2023 Marsh J. Ray
//
// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
// http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
// http://opensource.org/licenses/MIT>, at your option. This file may not be
// copied, modified, or distributed except according to those terms.

#![allow(dead_code)] //? TODO for development

use std::io::{BufRead, Read};

use serde::{Deserialize, Serialize};

#[derive(Debug, PartialEq, Eq, Deserialize, Serialize)]
pub enum ByteOrEof {
    Byte(u8),
    StdIoError(String),
    Eof,
}

#[derive(Debug, PartialEq, Eq, Deserialize, Serialize)]
pub struct SourceByteReadResult {
    pub file_offset: u64,
    pub byte_or_eof: ByteOrEof,
}

#[allow(dead_code)]
pub fn source_bytes(bx_bufread: Box<dyn BufRead>) -> impl Iterator<Item = SourceByteReadResult> {
    let bytes = bx_bufread.bytes();

    let bytes_eod = bytes.map(Some).chain(std::iter::once(None));

    bytes_eod.scan(0u64, |state, opt_stdioresult| {
        let file_offset = *state;
        *state += 1;

        Some(SourceByteReadResult {
            file_offset,
            byte_or_eof: match opt_stdioresult {
                Some(Result::Ok(by)) => ByteOrEof::Byte(by),
                Some(Result::Err(stdioerror)) => ByteOrEof::StdIoError(stdioerror.to_string()),
                None => ByteOrEof::Eof,
            },
        })
    })
}

#[cfg(test)]
mod test {
    #[test]
    fn test() {
        const TEST_DATA_SUBDIR: &str = "source_bytes";

        crate::test_util::insta_glob(TEST_DATA_SUBDIR, |_file_path, bx_bufread| {
            let source_bytes = crate::source_bytes::source_bytes(bx_bufread);

            let results = source_bytes.collect::<Vec<_>>();

            insta::assert_ron_snapshot!(results);
        });
    }
}
