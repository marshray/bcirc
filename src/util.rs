// Copyright 2023 Marsh J. Ray
//
// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
// http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
// http://opensource.org/licenses/MIT>, at your option. This file may not be
// copied, modified, or distributed except according to those terms.

#![allow(dead_code)] //? TODO for development
#![allow(unused_mut)] //? TODO for development
#![allow(unused_variables)] //? TODO for development
#![allow(unused_imports)] //? TODO for development

/// Build an Option<> from a conditional and an expression.
///
/// # Examples
///
/// ```
/// # use std::ops::Range;
/// # use bcirc::maybe_some;
/// struct X(usize);
/// impl X {
///     const VALID_RANGE: Range<usize> = 1_usize .. 101;
///     fn opt_x_from_usize(val: usize) -> Option<X> {
///         maybe_some!(Self::VALID_RANGE.contains(&val), X(val))
///     }
/// }
/// assert!(X::opt_x_from_usize(100).is_some());
/// assert!(X::opt_x_from_usize(101).is_none());
/// ```
///
#[macro_export]
macro_rules! maybe_some {
    ($e:expr, $f:expr) => {
        if $e {
            std::option::Option::Some($f)
        } else {
            std::option::Option::None
        }
    };
}

/// Like `std::matches`, but allows multiple patterns.
///
/// # Examples
///
/// ```
/// # use bcirc::matches_any;
/// fn get_message(chars: &[char; 2]) -> bool {
///     matches_any!(chars,
///         &[ a, _b ] if a == 'E',
///         &[ _a, b ] if b == 'H' )
/// }
/// assert_eq!(get_message(&['E', 'F']), true);
/// assert_eq!(get_message(&['F', 'G']), false);
/// assert_eq!(get_message(&['G', 'H']), true);
/// ```
#[macro_export]
macro_rules! matches_any {
    ($ex:expr, $( $pa:pat $(if $gu:expr)? ),*) => {
        match $ex {
            $( $pa $(if $gu)? => true, )*
            _ => false
        }
    };
}

/// Build an Option<> from a match expression.
///
/// # Examples
///
/// ```
/// # use bcirc::maybe_match;
/// struct CharStruct { ch: char };
/// let char_struct = CharStruct { ch: 'F' };
/// let opt_ch: Option<char> = maybe_match!(
///     char_struct, CharStruct { ch, .. } if ch != '\0' => ch
/// );
/// assert!(opt_ch.is_some());
/// ```
#[macro_export]
macro_rules! maybe_match {
    ($scrutinee:expr, $patt:pat $( if $guard:expr )? => $some_expr:expr) => {
        match $scrutinee {
            $patt $(if $guard)? => std::option::Option::Some($some_expr),
            _ => std::option::Option::None
        }
    };
}

/// A trait for fallible conversions when no Error type is specified.
/// TODO rename MaybeFrom?
pub trait OptFrom<T>: Sized {
    fn opt_from(value: T) -> Option<Self>;

    /// Produce a `std::result::Result<T, &'static str>` with a generic error message.
    fn try_from_ok_or_err(val: T) -> Result<Self, &'static str> {
        Self::opt_from(val).ok_or("Couldn't convert val.")
    }
}

/* Alas.

error[E0210]: type parameter `T` must be used as the type parameter for some local type (e.g., `MyStruct<T>`)
   = note: implementing a foreign trait is only possible if at least one of the types for which it is implemented is local
   = note: only traits defined in the current crate can be implemented for a type parameter

impl<T: OptFrom<U>, U> std::convert::TryFrom<U> for T
where
    T: OptFrom<U>,
{
    type Error = &'static str;

    fn try_from(val: U) -> Result<Self, Self::Error> {
        val.opt_from(val).ok_or(Err("OptFrom couldn't convert."))
    }
} */

pub(crate) const fn one_shl(ch: char) -> u128 {
    let ch = ch as u32;
    if ch < 128 {
        1_u128 << ch
    } else {
        assert!(ch < 128);
        0
    }
}

pub(crate) const fn fs_shl(n: u32, ch: char) -> u128 {
    let ch = ch as u32;
    assert!(n + ch < 128);
    ((1_u128 << n) - 1) << ch
}

pub(crate) const fn u128_ch_bit_test(u: u128, ch: char) -> bool {
    let ch = ch as u32;
    ch < 128 && ((u >> ch) & 1) != 0
}
