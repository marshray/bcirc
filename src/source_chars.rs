// Copyright 2023 Marsh J. Ray
//
// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
// http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
// http://opensource.org/licenses/MIT>, at your option. This file may not be
// copied, modified, or distributed except according to those terms.

use std::ops::RangeInclusive;

use serde::{Deserialize, Serialize};

use crate::maybe_match;
use crate::source_bytes::{ByteOrEof, SourceByteReadResult};
use crate::util::{one_shl, u128_ch_bit_test};

#[repr(transparent)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Deserialize, Serialize)]
pub struct LineNumberOneBased(u64);

impl LineNumberOneBased {
    #[allow(dead_code)] // I don't know why it doesn't pick up the usage below.
    pub fn inc(&mut self) {
        self.0 += 1;
    }
}

#[repr(transparent)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Deserialize, Serialize)]
pub struct CharNumberOneBased(u64);

impl CharNumberOneBased {
    #[allow(dead_code)] // I don't know why it doesn't pick up the usage below.
    pub fn inc(&mut self) {
        self.0 += 1;
    }
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct CharLoc {
    pub fo_range: RangeInclusive<u64>,
    pub line_n: LineNumberOneBased,
    pub char_n: CharNumberOneBased,
}

impl CharLoc {
    pub fn new() -> CharLoc {
        CharLoc {
            fo_range: 0u64..=0,
            line_n: LineNumberOneBased(1),
            char_n: CharNumberOneBased(1),
        }
    }
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub enum CharError {
    InvalidUtf8(u8),
    UnexpectedEof,
    StdIoError(String),
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub enum CharResult {
    Char(char),
    Eol,
    Eof,
    CharError(CharError),
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct SourceCharReadResult {
    pub loc: CharLoc,
    pub char_result: CharResult,
}

impl SourceCharReadResult {
    pub fn default_error() -> Self {
        Self {
            loc: CharLoc::new(),
            char_result: CharResult::Eof,
        }
    }

    /// Returns if the result is a valid Eol.
    pub fn is_eol(&self) -> bool {
        matches!(
            *self,
            SourceCharReadResult {
                char_result: CharResult::Eol,
                ..
            }
        )
    }

    /// Returns if the result is a valid Eof.
    pub fn is_eof(&self) -> bool {
        matches!(
            *self,
            SourceCharReadResult {
                char_result: CharResult::Eof,
                ..
            }
        )
    }

    /// Returns if the result is an error.
    pub fn is_error(&self) -> bool {
        matches!(
            *self,
            SourceCharReadResult {
                char_result: CharResult::CharError(_),
                ..
            }
        )
    }

    /// Returns if the result is a char.
    pub fn is_char(&self) -> bool {
        matches!(
            *self,
            SourceCharReadResult {
                char_result: CharResult::Char(_),
                ..
            }
        )
    }

    /// Returns if the result is a char or a valid Eol.
    pub fn is_char_or_eol(&self) -> bool {
        self.is_char() || self.is_eol()
    }

    /// Returns the char, if the SourceCharReadResult contains one.
    pub fn opt_char(&self) -> Option<char> {
        maybe_match!(*self, SourceCharReadResult { char_result: CharResult::Char(ch), .. } => ch)
    }

    /// Returns the char if the SourceCharReadResult contains one. Otherwise returns
    /// NUL (0 as char).
    pub fn char_or_nul(&self) -> char {
        self.opt_char().unwrap_or('\0')
    }
}

const fn is_cr_or_lf(ch: char) -> bool {
    const PUNCTUATION_FIRST_CHAR: u128 = one_shl('\r') | one_shl('\n');

    u128_ch_bit_test(PUNCTUATION_FIRST_CHAR, ch)
}

#[allow(dead_code)] //? TODO Hard to get this to be recognized as used for some reason.
pub fn source_chars(
    source_bytes: &mut dyn Iterator<Item = SourceByteReadResult>,
) -> impl Iterator<Item = SourceCharReadResult> + '_ {
    let mut source_char_read_result = SourceCharReadResult {
        loc: CharLoc::new(),
        char_result: CharResult::Eof,
    };

    let mut utf8_buf = [0u8; 4];
    let mut utf8_buf_len = 0;
    let mut need_more_bytes_for_this_char = false;
    let mut track_prev_char_was_cr = false;

    std::iter::from_generator(move || {
        for source_byte_read_result in source_bytes {
            // Figure out the file offset range of the new char.
            source_char_read_result.loc.fo_range = {
                let byte_fo = source_byte_read_result.file_offset;
                let mut char_fo_start = *source_char_read_result.loc.fo_range.start();

                if need_more_bytes_for_this_char {
                    // Just extend existing char
                    need_more_bytes_for_this_char = false;
                } else {
                    // Begin a new char
                    utf8_buf_len = 0;
                    char_fo_start = byte_fo;
                }

                char_fo_start..=byte_fo
            };

            match source_byte_read_result.byte_or_eof {
                ByteOrEof::Byte(by) => {
                    assert!(utf8_buf_len < 4);
                    utf8_buf[utf8_buf_len] = by;
                    utf8_buf_len += 1;

                    match std::str::from_utf8(&utf8_buf[0..utf8_buf_len]) {
                        Result::Ok(ch_str) => {
                            assert_eq!(ch_str.len(), 1);
                            let ch = ch_str.chars().next().unwrap();

                            let prev_char_was_cr = track_prev_char_was_cr;
                            track_prev_char_was_cr = ch == '\r';

                            if prev_char_was_cr && ch == '\n' {
                                // Ignore LF after CR
                            } else {
                                let yielding_eol = is_cr_or_lf(ch);

                                let char_result = if yielding_eol {
                                    CharResult::Eol
                                } else {
                                    CharResult::Char(ch)
                                };

                                let scrr = SourceCharReadResult {
                                    char_result,
                                    ..source_char_read_result.clone()
                                };
                                //eprintln!("\nyielding {:?}", &scrr);

                                yield scrr;

                                if yielding_eol {
                                    //eprintln!("advancing to next line: {:?}", &_char_or_eof);
                                    // Advance to next line
                                    source_char_read_result.loc.char_n = CharNumberOneBased(1);
                                    source_char_read_result.loc.line_n.inc();
                                } else {
                                    //eprintln!("advancing to next char: {:?}", &_char_or_eof);
                                    // Advance to next char
                                    source_char_read_result.loc.char_n.inc();
                                }
                            }
                        }
                        Result::Err(utf8error) => {
                            match utf8error.error_len() {
                                None => {
                                    // Need more bytes for this char
                                    need_more_bytes_for_this_char = true;
                                    // continue
                                }
                                Some(_len) => {
                                    // An invalid Utf8 byte was encountered
                                    debug_assert_eq!(utf8error.valid_up_to(), 0);

                                    let char_error = CharError::InvalidUtf8(by);
                                    yield SourceCharReadResult {
                                        char_result: CharResult::CharError(char_error),
                                        ..source_char_read_result.clone()
                                    };
                                }
                            } // match
                        } // Err
                    } // match from_utf8
                } // ByteOrEof::Byte
                ByteOrEof::StdIoError(err_str) => {
                    yield SourceCharReadResult {
                        char_result: CharResult::CharError(CharError::StdIoError(err_str)),
                        ..source_char_read_result.clone()
                    };
                }
                ByteOrEof::Eof => {
                    if utf8_buf_len != 0 {
                        yield SourceCharReadResult {
                            char_result: CharResult::CharError(CharError::UnexpectedEof),
                            ..source_char_read_result.clone()
                        };
                    } else {
                        if source_char_read_result.loc.char_n != CharNumberOneBased(1) {
                            yield SourceCharReadResult {
                                char_result: CharResult::Eol,
                                ..source_char_read_result.clone()
                            };
                            source_char_read_result.loc.char_n.inc();
                        }

                        yield SourceCharReadResult {
                            char_result: CharResult::Eof,
                            ..source_char_read_result.clone()
                        };
                    };
                }
            } // match
        } // for in source_bytes
    }) // from_generator
}

#[cfg(test)]
mod test {
    #[test]
    fn test() {
        const TEST_DATA_SUBDIR: &str = "source_chars";

        crate::test_util::insta_glob(TEST_DATA_SUBDIR, |_file_path, bx_bufread| {
            let mut source_bytes = crate::source_bytes::source_bytes(bx_bufread);
            let source_chars = crate::source_chars::source_chars(&mut source_bytes);

            let results = source_chars.collect::<Vec<_>>();

            insta::assert_ron_snapshot!(results);
        });
    }
}
