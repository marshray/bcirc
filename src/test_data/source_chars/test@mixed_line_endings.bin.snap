---
source: src/source_chars.rs
input_file: src/test_data/source_chars/mixed_line_endings.bin
---
[
  SourceCharReadResult(
    loc: CharLoc(
      fo_range: RangeInclusive(
        start: 0,
        end: 0,
      ),
      line_n: LineNumberOneBased(1),
      char_n: CharNumberOneBased(1),
    ),
    char_result: Char('a'),
  ),
  SourceCharReadResult(
    loc: CharLoc(
      fo_range: RangeInclusive(
        start: 1,
        end: 1,
      ),
      line_n: LineNumberOneBased(1),
      char_n: CharNumberOneBased(2),
    ),
    char_result: Eol,
  ),
  SourceCharReadResult(
    loc: CharLoc(
      fo_range: RangeInclusive(
        start: 2,
        end: 2,
      ),
      line_n: LineNumberOneBased(2),
      char_n: CharNumberOneBased(1),
    ),
    char_result: Char('b'),
  ),
  SourceCharReadResult(
    loc: CharLoc(
      fo_range: RangeInclusive(
        start: 3,
        end: 3,
      ),
      line_n: LineNumberOneBased(2),
      char_n: CharNumberOneBased(2),
    ),
    char_result: Eol,
  ),
  SourceCharReadResult(
    loc: CharLoc(
      fo_range: RangeInclusive(
        start: 5,
        end: 5,
      ),
      line_n: LineNumberOneBased(3),
      char_n: CharNumberOneBased(1),
    ),
    char_result: Char('c'),
  ),
  SourceCharReadResult(
    loc: CharLoc(
      fo_range: RangeInclusive(
        start: 6,
        end: 6,
      ),
      line_n: LineNumberOneBased(3),
      char_n: CharNumberOneBased(2),
    ),
    char_result: Eol,
  ),
  SourceCharReadResult(
    loc: CharLoc(
      fo_range: RangeInclusive(
        start: 7,
        end: 7,
      ),
      line_n: LineNumberOneBased(4),
      char_n: CharNumberOneBased(1),
    ),
    char_result: Eol,
  ),
  SourceCharReadResult(
    loc: CharLoc(
      fo_range: RangeInclusive(
        start: 9,
        end: 9,
      ),
      line_n: LineNumberOneBased(5),
      char_n: CharNumberOneBased(1),
    ),
    char_result: Char('d'),
  ),
  SourceCharReadResult(
    loc: CharLoc(
      fo_range: RangeInclusive(
        start: 10,
        end: 10,
      ),
      line_n: LineNumberOneBased(5),
      char_n: CharNumberOneBased(2),
    ),
    char_result: Eol,
  ),
  SourceCharReadResult(
    loc: CharLoc(
      fo_range: RangeInclusive(
        start: 11,
        end: 11,
      ),
      line_n: LineNumberOneBased(6),
      char_n: CharNumberOneBased(1),
    ),
    char_result: Eol,
  ),
  SourceCharReadResult(
    loc: CharLoc(
      fo_range: RangeInclusive(
        start: 12,
        end: 12,
      ),
      line_n: LineNumberOneBased(7),
      char_n: CharNumberOneBased(1),
    ),
    char_result: Char('e'),
  ),
  SourceCharReadResult(
    loc: CharLoc(
      fo_range: RangeInclusive(
        start: 13,
        end: 13,
      ),
      line_n: LineNumberOneBased(7),
      char_n: CharNumberOneBased(2),
    ),
    char_result: Eol,
  ),
  SourceCharReadResult(
    loc: CharLoc(
      fo_range: RangeInclusive(
        start: 13,
        end: 13,
      ),
      line_n: LineNumberOneBased(7),
      char_n: CharNumberOneBased(3),
    ),
    char_result: Eof,
  ),
]
